lib/puppet/data_providers/hiera_interpolate.rb

```diff
@@ -100,10 +100,16 @@ def qualified_lookup(segments, value)
       throw :no_such_key if value.nil?
       if segment =~ /^[0-9]+$/
         segment = segment.to_i
-        raise Puppet::DataBinding::LookupError, "Data provider type mismatch: Got #{value.class.name} when Array was expected to enable lookup using key '#{segment}'" unless value.instance_of?(Array)
+        unless value.instance_of?(Array)
+          raise Puppet::DataBinding::LookupError,
+                "Data provider type mismatch: Got #{value.class.name} when Array was expected to enable lookup using key '#{segment}'"
+        end
         throw :no_such_key unless segment < value.size
       else
-        raise Puppet::DataBinding::LookupError, "Data provider type mismatch: Got #{value.class.name} when a non Array object that responds to '[]' was expected to enable lookup using key '#{segment}'" unless value.respond_to?(:'[]') && !value.instance_of?(Array)
+        unless value.respond_to?(:'[]') && !(value.instance_of?(Array) || value.instance_of?(String))
```

Comment:
I don't think 80 characters is a good limit. Perhaps in commit comments, but not in code. I tend to use a 120 character limit.